src/v/kafka/protocol/schemata/generator.py

```diff
@@ -1128,6 +1148,51 @@ def is_error_code(self):
 {%- endif %}
 {% endmacro %}
 
+{% macro render_formatter(struct) %}
+{%- if not struct.is_streamable %}
+template<>
+struct fmt::formatter<kafka::{{ struct.name }}, char> {
+
+    template <typename ParseContext>
+    constexpr auto parse(ParseContext &ctx) -> decltype(ctx.begin()) {
+        return ctx.begin();
+    }
+
+    template <typename FormatContext>
+    auto format([[maybe_unused]] const kafka::{{ struct.name }} &v, FormatContext& ctx) const
+            -> decltype(ctx.out()) {
+        return fmt::format_to(ctx.out(),
+          "{{'{{' + struct.format + '}}'}}",
+          {%- for field in struct.fields %}
+          {%- if field.is_sensitive %}"****"{% else %}v.{{ field.name }}{% endif %}{% if not loop.last %},{% endif %}
+
+          {%- endfor %}
```

Comment:
it looks like this goes into the header? if so, can we put it into the generated .cc file?

Reply:
Yeah, that would be better...been fighting the linker

Reply:
done